src/client/auth/scram.rs

```diff
@@ -92,15 +105,61 @@ impl ScramVersion {
             ));
         };
 
-        let nonce = auth::generate_nonce();
+        Ok(ClientAuthInfo {
+            username,
+            password,
+            source,
+        })
+    }
+
+    /// Constructs the first client message in the SCRAM handshake.
+    pub(crate) fn build_client_first(&self, credential: &Credential) -> Result<ClientFirst> {
+        let info = self.client_auth_info(credential)?;
 
-        let client_first = ClientFirst::new(source, username, nonce.as_str());
+        Ok(ClientFirst::new(info.source, info.username))
+    }
+
+    /// Sends the first client message in the SCRAM handshake.
+    async fn send_client_first(
+        &self,
+        conn: &mut Connection,
+        credential: &Credential,
+    ) -> Result<FirstRound> {
+        let client_first = self.build_client_first(credential)?;
 
         let command = client_first.to_command(self);
 
-        let server_first_response = conn.send_command(command, None).await?;
-        let server_first = ServerFirst::parse(server_first_response.raw_response)?;
-        server_first.validate(nonce.as_str())?;
+        let server_first = conn.send_command(command, None).await?;
+
+        Ok(FirstRound {
+            client_first,
+            server_first,
+        })
+    }
+
+    /// Perform SCRAM authentication for a given stream.
+    pub(crate) async fn authenticate_stream(
+        &self,
+        conn: &mut Connection,
+        credential: &Credential,
+        first_round: impl Into<Option<FirstRound>>,
```

Comment:
I updated the handshakes to optionally take a FirstRound struct that contained both the first client message (which was sent as part of the handshake) as well as the server's response to the speculative authentication attempt. If this is present, the handshake will just skip the first round trip and proceed in the same fashion as it would if it had performed the handshake normally.